cpp/include/cudf/io/datasource.hpp

```diff
@@ -0,0 +1,182 @@
+   */
+  static std::unique_ptr<datasource> create(datasource* source);
+
+  /**
+   * @brief Base class destructor
+   */
+  virtual ~datasource(){};
+
+  /**
+   * @brief Returns a buffer with a subset of data from the source.
+   *
+   * @param[in] offset Bytes from the start
+   * @param[in] size Bytes to read
+   *
+   * @return The data buffer
+   */
+  virtual std::unique_ptr<datasource::buffer> host_read(size_t offset, size_t size) = 0;
+
+  /**
+   * @brief Reads a selected range into a preallocated buffer.
+   *
+   * @param[in] offset Bytes from the start
+   * @param[in] size Bytes to read
+   * @param[in] dst Address of the existing host memory
+   *
+   * @return The number of bytes read (can be smaller than size)
+   */
+  virtual size_t host_read(size_t offset, size_t size, uint8_t* dst) = 0;
+
+  /**
+   * @brief Whether or not this source supports reading directly into device memory.
+   *
+   * If this function returns true, the datasource will receive calls to device_read() instead of
+   * host_read() when possible. However, it is still possible to receive host_read() calls as well.
+   *
+   * Data source implementations that don't support direct device reads don't need to override this
+   * function. The implementations that do should override it to return false.
+   *
+   * @return bool Whether this source supports device_read() calls
+   */
+  virtual bool supports_device_read() const { return false; }
+
+  /**
+   * @brief Returns a device buffer with a subset of data from the source.
+   *
+   * Data source implementations that don't support direct device reads don't need to override this
+   * function.
+   *
+   * @param[in] offset Bytes from the start
+   * @param[in] size Bytes to read
+   *
+   * @return The data buffer in the device memory
+   */
+  virtual std::unique_ptr<datasource::buffer> device_read(size_t offset, size_t size)
+  {
+    CUDF_FAIL("datasource classes that support device_read must override this function.");
+  }
+
+  /**
+   * @brief Reads a selected range into a preallocated device buffer
+   *
+   * Data source implementations that don't support direct device reads don't need to override this
+   * function.
+   *
+   * @param[in] offset Bytes from the start
+   * @param[in] size Bytes to read
+   * @param[in] dst Address of the existing device memory
+   *
+   * @return The number of bytes read (can be smaller than size)
+   */
+  virtual size_t device_read(size_t offset, size_t size, uint8_t* dst)
+  {
+    CUDF_FAIL("datasource classes that support device_read must override this function.");
+  }
+
+  /**
+   * @brief Returns the size of the data in the source.
+   *
+   * @return size_t The size of the source data in bytes
+   */
+  virtual size_t size() const = 0;
+
+  /**
+   * @brief Returns whether the source contains any data.
+   *
+   * @return bool True if there is data, False otherwise
+   */
+  virtual bool empty() const { return size() == 0; }
```

Comment:
I've been told that best practice in C++ for  this type of function is to prefix it with `is`. `std::vector::empty()` is apparently considered a mistake.

```suggestion
  virtual bool is_empty() const { return size() == 0; }
```